### DISCUSSION TITLE
Severity HIGH security problem to be announced with curl 8.4.0 on Oct 11

### DISCUSSION THREAD

**Discussion by bagder** [curl/curl#12026] [General]
We are cutting the release cycle short and will release [curl 8.4.0](https://daniel.haxx.se/blog/2023/10/11/curl-8-4-0/) on **October 11**, including fixes for a severity HIGH CVE and one severity LOW. The one rated HIGH is probably the worst curl security flaw in a long time.

The new version and details about the two CVEs will be published around 06:00 UTC on the release day.

- [CVE-2023-38545](https://curl.se/docs/CVE-2023-38545.html): severity HIGH (affects both libcurl and the curl tool)
- [CVE-2023-38546](https://curl.se/docs/CVE-2023-38546.html): severity LOW (affects libcurl only, not the tool)

There is no API nor ABI change in the coming curl release.

I cannot disclose any information about which version range that is affected, as that would help identify the problem (area) with a very high accuracy so I cannot do that ahead of time. The "last several years" of versions is as specific as I can get.

We have notified the [distros mailing list](https://oss-security.openwall.org/wiki/mailing-lists/distros) allowing the member distributions to prepare patches. (**No one** else gets details about these problems before October 11 without a support contract and a good reason.)

Now you know. Plan accordingly.

Bonus: [How I made a heap overflow in curl](https://daniel.haxx.se/blog/2023/10/11/how-i-made-a-heap-overflow-in-curl/)

**Comment by dreirund** [comment]
> *release curl 8.4.0 on October 11*
> *[...]*
> *probably the worst curl security flaw in a long time.*

When there is a HIGH CVE security flaw, why then not release immediately after fix has been applied, but at a set date? (Now, exploit-hunters are aware that there is something and have 7 days notice in which they can focus intensely on curl to find the flaw and exploit it.)

**Reply by tobixen** [reply]
As soon as the fix has been published, exploiters have all the information they need to abuse the flaw, so then it's important that OS-distributors and users of the code are prepared to distribute and patch ASAP.  Hence the importance of a preparation window.

Addendum: it's also important to spend some time on quality assurance.  Every now and then we (system administrators) are rushing to patch our systems due to some really nasty security problem, and as soon as we're done we learn that the hastily developed bugfix has introduced new security problems and/or new bugs ... I don't remember which bug that was (possibly OpenSSL or log4j), but on the worst there were three urgent patch rounds on the same software component within a week.

**Reply by bagder** [reply]
I set the date to
- allow us a few days for more deliberating on the vulnerability, to really think it through, write the advisory, understand it proper. Rinse and repeat.
- give "distro people" a few days to prepare patched updates
- allow a few days for the project (and me) to line up things to prepare for the new release
- we can spread the word about the pending release and the main reason for it in the mean time
- the release needs to work with my personal schedule and Wednesdays are our standard release days

Sure, there is a minuscule risk that someone can find this (again) before we ship the patch, but this issue has stayed undetected for years for a reason. I think taking a few days to make sure we do a solid release is worth this risk.

**Reply by realsimix** [reply]
I'm wondering about the term "distro people": who is this exactly and how do they handle this case?

I guess that they already know more than we, the public, do. How can we be sure they will fix this issue in time and make sure those doing derivative work will also follow in time?

Let's say, if Red Hat can already prepare updates because they have access to some still secret information, how can the creators of all its clones, like Rocky, Alma, Oracle, CensOS and the like also provide updates in time?

**Reply by ThiefMaster** [reply]
Usually trusted people from mainstream Linux distributions have the proper contacts (and/or are on the proper private mailing lists) to receive patches and more information than the general public early enough. I can only assume that this is the case here as well....

**Reply by bagder** [reply]
The "distro people" I refer to above are the ones subscribed to the [distros](https://oss-security.openwall.org/wiki/mailing-lists/distros) mailing list.

**Reply by stevenpackardblp** [reply]
Being one of those "distro people" maintaining an internal distribution for my company only, it would be beneficial if we could somehow get access to a patch that can be applied to the 8.3.0 source as soon as possible.

**Comment by jbeckers** [comment]
Any guidance on which versions are vulnerable, so we can start planning what to upgrade?

**Reply by bagder** [reply]
I'm sorry, but if I would give you that information it would help identify the problem (area) with a very high accuracy so I cannot do that ahead of time. The "last several years" of versions is as specific as I can get.

**Reply by robertsdotpm** [reply]
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢀⡀⠴⠤⠤⠴⠄⡄⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⣠⠄⠒⠉⠀⠀⠀⠀⠀⠀⠀⠀⠁⠃⠆⡄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⢀⡜⠁⠀⠀⠀⢠⡄⠀⣀⠀⠀⠀⠀⠀⠀⠀⠀⠑⡄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⢈⠁⠀⠀⠠⣿⠿⡟⣀⡹⠆⡿⣃⣰⣆⣤⣀⠀⠀⠹⡄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⣼⠀⠀⢀⣀⣀⣀⣀⡈⠁⠙⠁⠘⠃⠡⠽⡵⢚⠱⠂⠛⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠈⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⡆⠀⠀⠀⠀⢐⣢⣤⣵⡄⢀⠀⢀⢈⣉⠉⠉⠒⠤⠀⠿⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠘⡇⠀⠀⠀⠀⠀⠉⠉⠁⠁⠈⠀⠸⢖⣿⣿⣷⠀⠀⢰⡆⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⡇⠀⠀⠀⠀⠀⠀⠀⠀⢀⠃⠀⡄⠀⠈⠉⠀⠀⠀⢴⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⢈⣇⠀⠀⠀⠀⠀⠀⠀⢰⠉⠀⠀⠱⠀⠀⠀⠀⠀⢠⡄⠀⠀⠀⠀⠀⣀⠔⠒⢒⡩⠃⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⣴⣿⣤⢀⠀⠀⠀⠀⠀⠈⠓⠒⠢⠔⠀⠀⠀⠀⠀⣶⠤⠄⠒⠒⠉⠁⠀⠀⠀⢸⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⡄⠤⠒⠈⠈⣿⣿⣽⣦⠀⢀⢀⠰⢰⣀⣲⣿⡐⣤⠀⠀⢠⡾⠃⠀⠀⠀⠀⠀⠀⠀⣀⡄⣠⣵⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠘⠏⢿⣿⡁⢐⠶⠈⣰⣿⣿⣿⣿⣷⢈⣣⢰⡞⠀⠀⠀⠀⠀⠀⢀⡴⠋⠁⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠈⢿⣿⣍⠀⠀⠸⣿⣿⣿⣿⠃⢈⣿⡎⠁⠀⠀⠀⠀⣠⠞⠉⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠈⢙⣿⣆⠀⠀⠈⠛⠛⢋⢰⡼⠁⠁⠀⠀⠀⢀⠔⠁⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠚⣷⣧⣷⣤⡶⠎⠛⠁⠀⠀⠀⢀⡤⠊⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠁⠈⠁⠀⠀⠀⠀⠀⠠⠊⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠃⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠸⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀

**Comment by taladar** [comment]
I know you can't give any details before the fix is released but I assume from the way the post is phrased that you are not talking about a HIGH security flaw in some obscure protocol or similar corner of the code virtually nobody ever uses in a way that is exploitable but about something that would affect a significant percentage of the curl and libcurl user bases?

**Reply by bagder** [reply]
Every security flaw requires a set of conditions to apply for the problem to trigger. The pending security vulnerabilities are no different. I cannot comment on what that set is ahead of time.

The severity level is a blunt tool. This is a HIGH severity problem but there is still going to be a large chunk of users who will not be affected by these problems.

**Comment by conmurray** [comment]
@bagder Might I ask if you think this would also affect pycurl, python-pycurl etc?

Thanks in advance

Conor

**Reply by bagder** [reply]
Yes it will. In general terms: *everything* that uses libcurl could theoretically use libcurl in a way that triggers this vulnerability, assuming that the conditions apply and that a vulnerable libcurl version is used. Of course some/many users will also use libcurl *without* being able to trigger the vulnerability.

It is impossible for me to make affirmative statements about specific libcurl users now.

**Reply by davidfoerster** [reply]
In case you’re asking if the *patch* to libcurl will require a redistribution of pycurl: most distros ship pycurl with dynamic linkage of libcurl (rather than static linkage). In that case, pycurl would *almost certainly* benefit from the changes to the (dynamically linked) libcurl without repackaging/redistribution.

(I say “almost certainly” because the libcurl patch might introduce a relevant ABI change which would require changes to pycurl but I find that very unlikely since there’s no such notice in the OP above.)

**Reply by bagder** [reply]
There is no API nor ABI change in the coming curl release. Updating the shared libcurl library should be enough to fix this issue on all operating systems.

Then again there will also be countless docker (and similar) images that feature their own copies, so there will still be quite a large number of rebuilds necessary I bet.

**Comment by robertsdotpm** [comment]
I really don't know why you posted this. You: 'theres a super bad bug in curl!!!!! OMGG OMGGGGG' and also you: 'no, i wotn give u any details, lol sry, bye.' This is how you cause the most damage possible for no reason. The correct way to handle this would have been to privately notify relevant stakeholders (maintainers for repos and the curl team.) This is just max panic mode with a half-assed notice that helps exploiters more than anything. Also, lmao at the downvotes you soy fucks.

**Reply by elfrinjo** [reply]
Well, it helps me to allocate resources for patching, testing and containerbuilding in the next week

**Reply by 19h** [reply]
@robertsdotpm there's literally zero information given about the locality of the vulnerability, therefore there's no knowledge asymmetry. This announcement serves to prepare admins for action when the fixes are ready for immediate release as opposed to "look for the fix on your system some unspecified time in the future."

**Reply by Woovie** [reply]
Go back to 4chan

**Reply by Rattenhirn** [reply]
@robertsdotpm I really don't know why you posted this.

**Reply by robertsdotpm** [reply]
Sorry that I have to spoon feed you morons what responsible disclosure is. This announcement is idiotic and unprecedented.

**Reply by ThiefMaster** [reply]
Bullshit. OpenSSL did the same for example: https://mta.openssl.org/pipermail/openssl-announce/2022-October/000238.html
Now go troll elsewhere.

**Reply by robertsdotpm** [reply]
Wow man, this one software project really proves your point here. As opposed to my approach widely used by every major software company and the entire security industry. You were also sure to select the perfect exemplar, OpenSSL, widely known for its amazing usability and high cryptographic security. (Wait) (Wait?) (OH.) OpenSSL is a steaming pile of shit and you're still wrong.

**Reply by ThiefMaster** [reply]
Well "this one software project" is one of the most critical ones out there... And I think we all know that OpenSSL is far from perfect (even though I think they improved a lot over the last years) - but guess what ,everyone was happy that they announced it in advance so people could prepare to patch ASAP when the patch hit.

Do you seriously prefer to be surprised - maybe on your day off, away from a proper computer - with an urgent update you need to install, instead of knowing some days in advance that you may need to quickly install a patch on a given day?

**Reply by bagder** [reply]
@robertsdotpm if you cannot keep a good and productive tone in this discussion from now on, I will ban you from this repository

**Reply by robertsdotpm** [reply]
⠀⠀⠘⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⡜
⠀⠀⠀⠑⡀⠀                   ⡔⠁
⠀⠀⠀⠀⠈⠢⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣀⠴⠊
⠀⠀⠀⠀⠀⠀⠀⢸⠀⠀⠀⢀⣀⣀⣀⣀⣀⡀⠤⠄⠒⠈
⠀⠀⠀⠀⠀⠀⠀⠘⣀⠄⠊⠁
⠀⠀⠀⠀⠀⠀⠀⠀⠀⢀⡠⠔⠒⠒⠒⠒⠒⠢⠤⣀
⠀⠀⠀⠀⠀⠀⠀⡰⠉⠁⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠑⢄⡀
⠀⠀⠀⠀⠀⠀⡸⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⡀⠀⠀⠀⠀⠙⠄
⠀⠀⠀⠀⠀⢀⠁⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠃⠀⢠⠂⠀⠀⠘⡄
⠀⠀⠀⠀⠀⢸⠀⠀⠀⠀⠀⠀⠀⠀⠈⢤⡀⢂⠀⢨⠀⢀⡠⠈⢣
⠀⠀⠀⠀⠀⢀⢀⡖⠒⠶⠤⠭⢽⣟⣗⠲⠖⠺⣖⣴⣆⡤⠤⠤⠼⡄
⠀⠀⠀⠀⠀⠘⡈⠃⠀⠀⠀⠘⣺⡟⢻⠻⡆⠀⡏⠀⡸⣿⢿⢞⠄⡇
⠀⠀⠀⠀⠀⠀⢣⡀⠤⡀⡀⡔⠉⣏⡿⠛⠓⠊⠁⠀⢎⠛⡗⡗⢳⡏
⠀⠀⠀⠀⠀⠀⠀⢱⠀⠨⡇⠃⠀⢻⠁⡔⢡⠒⢀⠀⠀⡅⢹⣿⢨⠇
⠀⠀⠀⠀⠀⠀⠀⢸⠀⠠⢼⠀⠀⡎⡜⠒⢀⠭⡖⡤⢭⣱⢸⢙⠆
⠀⠀⠀⠀⠀⠀⠀⡸⠀⠀⠸⢁⡀⠿⠈⠂⣿⣿⣿⣿⣿⡏⡍⡏
⠀⠀⠀⠀⠀⠀⢀⠇⠀⠀⠀⠀⠸⢢⣫⢀⠘⣿⣿⡿⠏⣼⡏
⠀⠀⠀⠀⣀⣠⠊⠀⣀⠎⠁⠀⠀⠀⠙⠳⢴⡦⡴⢶⣞⣁⣀⣀⡀
⠀⠐⠒⠉⠀⠀⠀⠀⠀⠀⠀⠀⠀⠠⠀⢀⠤⠀⠀⠀⠀⠀⠀⠀⠈⠉

**Comment by andreadelfino** [comment]
Thank you for the pre-alerting! At this stage can you disclose the affected versions range? It's just 8.* branch (given that you're going to release 8.4.0) or also 7.* or so may require additional patches? It would be helpful to better understand the impacted surface.

**Reply by ThiefMaster** [reply]
https://github.com/curl/curl/discussions/12026#discussioncomment-7194757

**Comment by bagder** [comment]
A lot of people are reading this thread or subscribe to it for real information updates, not suffering through troll comments. Therefore, **it is now locked** from further commenting. If you have additional questions/comments about the pending release and associated security vulnerabilities, contact me or start a separate discussion thread

I have tried to move relevant information into the initial post to reduce the need to scroll through this page to find it.

**Comment by bagder** [comment]
The release, the CVEs, the advisories and associated blog posts are now all public. The post here is updated with some links.
